amazon/ion/ioncmodule.c

```diff
@@ -384,48 +409,37 @@ static iERR ionc_write_sequence(hWRITER writer, PyObject* sequence, PyObject* tu
  */
 static iERR ionc_write_struct(hWRITER writer, PyObject* map, PyObject* tuple_as_sexp) {
     iENTER;
-    PyObject * list = PyMapping_Items(map);
-    PyObject * seq = PySequence_Fast(list, "expected a sequence within the map.");
-    PyObject * key = NULL, *val = NULL, *child_obj = NULL;
-    Py_ssize_t len = PySequence_Size(seq);
-    Py_ssize_t i;
-
-    for (i = 0; i < len; i++) {
-        child_obj = PySequence_Fast_GET_ITEM(seq, i);
-        key = PyTuple_GetItem(child_obj, 0);
-        val = PyTuple_GetItem(child_obj, 1);
-        Py_INCREF(child_obj);
-        Py_INCREF(key);
-        Py_INCREF(val);
-
-        if (PyUnicode_Check(key)) {
-            ION_STRING field_name;
-            ion_string_from_py(key, &field_name);
-            IONCHECK(ion_writer_write_field_name(writer, &field_name));
+    PyObject *store = NULL, *key = NULL, *val_list = NULL, *val = NULL;
+    Py_ssize_t pos = 0, i, list_len;
+    if (PyDict_Check(map)) {
+        while (PyDict_Next(map, &pos, &key, &val)) {
+            IONCHECK(write_struct_field(writer, key, val, tuple_as_sexp));
         }
-        else if (key == Py_None) {
-            // if field_name is None, write symbol $0 instead.
-            IONCHECK(_ion_writer_write_field_sid_helper(writer, 0));
+    } else {
+        store = PyObject_CallMethod(map, "_get_store", NULL);
+        if (store == NULL || !PyDict_Check(store)) {
+            IONCHECK(IERR_INVALID_ARG);
+            goto fail;
```

Comment:
Prefer `FAILWITH(IERR_INVALID_ARG)` or consider `_FAILWITHMSG(...)`.

Yes! I know I encouraged you to avoid `_` methods (in this case a macro) but for reasons I don't understand `_FAILWITHMSG` actually propagates the message back to the Python exception whereas the non-underscored doesn't.

Reply:
Dug into this a little..

`FAILWITHMSG` is provided by ion-c, but`_FAILWITHMSG` is defined in [ioncmodule.c](https://github.com/amazon-ion/ion-python/blob/master/amazon/ion/ioncmodule.c#L28), and copies the provided error msg into the global `_err_msg` . `ionc_read` and `ionc_write` (and other functions like iter read) use `_err_msg` when setting the error indicator (like [here](https://github.com/amazon-ion/ion-python/blob/master/amazon/ion/ioncmodule.c#L861)).

Reply:
Thanks for the suggestion. In this case I would choose to use  `_FAILWITHMSG(...)` to align with the err handling in other writing implementations.

Reply:
> Dug into this a little..
> 
> `FAILWITHMSG` is provided by ion-c, but`_FAILWITHMSG` is defined in [ioncmodule.c](https://github.com/amazon-ion/ion-python/blob/master/amazon/ion/ioncmodule.c#L28), and copies the provided error msg into the global `_err_msg` . `ionc_read` and `ionc_write` (and other functions like iter read) use `_err_msg` when setting the error indicator (like [here](https://github.com/amazon-ion/ion-python/blob/master/amazon/ion/ioncmodule.c#L861)).

Thanks for the information!